File: src/main/java/burp/AttackTableModel.java
package burp;

import javax.swing.table.AbstractTableModel;
import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class AttackTableModel extends AbstractTableModel {
    private List<AttackEntry> allAttacks;
    private List<AttackEntry> displayedAttacks;
    private final String[] columnNames = {"", "Time", "Category", "Method", "URL", "Status", "Tester"}; // Added empty column for checkbox
    private static final DateTimeFormatter DATE_FORMATTER = 
        DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss").withZone(ZoneId.systemDefault());
    
    private boolean showDeleted = false;
    private Set<String> selectedIds = new HashSet<>();

    public AttackTableModel() {
        this.allAttacks = new ArrayList<>();
        this.displayedAttacks = new ArrayList<>();
    }

    public void setAttacks(List<AttackEntry> attacks) {
        this.allAttacks = attacks;
        refreshDisplay();
    }

    public void setShowDeleted(boolean showDeleted) {
        this.showDeleted = showDeleted;
        refreshDisplay();
    }
    
    public boolean isShowDeleted() {
        return showDeleted;
    }

    private void refreshDisplay() {
        if (showDeleted) {
            displayedAttacks = allAttacks.stream()
                .filter(AttackEntry::isDeleted)
                .collect(Collectors.toList());
        } else {
            displayedAttacks = allAttacks.stream()
                .filter(a -> !a.isDeleted())
                .collect(Collectors.toList());
        }
        // Clear selection when switching views or refreshing data to avoid confusion
        selectedIds.clear(); 
        fireTableDataChanged();
    }

    public AttackEntry getAttackAt(int rowIndex) {
        if (rowIndex >= 0 && rowIndex < displayedAttacks.size()) {
            return displayedAttacks.get(rowIndex);
        }
        return null;
    }
    
    public List<AttackEntry> getDisplayedAttacks() {
        return displayedAttacks;
    }
    
    public Set<String> getSelectedIds() {
        return new HashSet<>(selectedIds);
    }
    
    public void setAllSelected(boolean selected) {
        if (selected) {
            for (AttackEntry attack : displayedAttacks) {
                selectedIds.add(attack.getId());
            }
        } else {
            selectedIds.clear();
        }
        fireTableDataChanged();
    }
    
    public boolean isAllSelected() {
        if (displayedAttacks.isEmpty()) return false;
        for (AttackEntry attack : displayedAttacks) {
            if (!selectedIds.contains(attack.getId())) {
                return false;
            }
        }
        return true;
    }

    public void clearSelection() {
        selectedIds.clear();
        fireTableDataChanged();
    }

    @Override
    public int getRowCount() {
        return displayedAttacks.size();
    }

    @Override
    public int getColumnCount() {
        return columnNames.length;
    }

    @Override
    public String getColumnName(int column) {
        return columnNames[column];
    }

    @Override
    public Class<?> getColumnClass(int columnIndex) {
        if (columnIndex == 0) {
            return Boolean.class;
        }
        return String.class;
    }

    @Override
    public boolean isCellEditable(int rowIndex, int columnIndex) {
        return columnIndex == 0;
    }

    @Override
    public Object getValueAt(int rowIndex, int columnIndex) {
        AttackEntry attack = displayedAttacks.get(rowIndex);
        switch (columnIndex) {
            case 0: return selectedIds.contains(attack.getId());
            case 1: return DATE_FORMATTER.format(Instant.ofEpochMilli(attack.getTimestamp()));
            case 2: return attack.getCategory();
            case 3: return attack.getMethod();
            case 4: return attack.getUrl();
            case 5: return attack.getStatus();
            case 6: return attack.getTesterName();
            default: return "";
        }
    }

    @Override
    public void setValueAt(Object aValue, int rowIndex, int columnIndex) {
        if (columnIndex == 0 && aValue instanceof Boolean) {
            AttackEntry attack = displayedAttacks.get(rowIndex);
            if ((Boolean) aValue) {
                selectedIds.add(attack.getId());
            } else {
                selectedIds.remove(attack.getId());
            }
            fireTableCellUpdated(rowIndex, columnIndex);
        }
    }
}

File: src/main/java/com/projectkepler/burp/AttackTableModel.kt
package com.projectkepler.burp

import javax.swing.table.AbstractTableModel
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

/**
 * Custom table model for displaying AttackEntry records.
 * Why: Using AbstractTableModel provides a reactive bridge between the data layer and Burp's Swing-based UI.
 * It allows us to handle complex selection states and filtering without blocking the UI thread.
 */
class AttackTableModel(private val storageManager: StorageManager) : AbstractTableModel() {
    private var allAttacks: MutableList<AttackEntry> = mutableListOf()
    var displayedAttacks: List<AttackEntry> = emptyList()
        private set

    private val columnNames = arrayOf("", "Time", "Category", "Method", "URL", "Status", "Tester")
    private val selectedIds = mutableSetOf<String>()

    var showDeleted: Boolean = false
        set(value) {
            field = value
            refreshDisplay()
        }

    /**
     * Updates the underlying data set.
     * Why: This triggers a full display refresh to ensure the UI stays in sync with the StorageManager.
     */
    fun setAttacks(attacks: List<AttackEntry>) {
        allAttacks = attacks.toMutableList()
        refreshDisplay()
    }

    /**
     * Filters the attacks based on deletion status and resets selection.
     * Why: Separating 'allAttacks' from 'displayedAttacks' allows us to toggle between
     * the active history and the trash without re-reading from disk.
     */
    private fun refreshDisplay() {
        displayedAttacks = allAttacks.filter { it.deleted == showDeleted }
        selectedIds.clear()
        fireTableDataChanged()
    }

    fun getAttackAt(rowIndex: Int): AttackEntry? = displayedAttacks.getOrNull(rowIndex)

    fun getSelectedIds(): Set<String> = selectedIds.toSet()

    fun setAllSelected(selected: Boolean) {
        if (selected) {
            displayedAttacks.forEach { selectedIds.add(it.id) }
        } else {
            selectedIds.clear()
        }
        fireTableDataChanged()
    }

    val isAllSelected: Boolean
        get() = displayedAttacks.isNotEmpty() && displayedAttacks.all { it.id in selectedIds }

    fun clearSelection() {
        selectedIds.clear()
        fireTableDataChanged()
    }

    override fun getRowCount(): Int = displayedAttacks.size

    override fun getColumnCount(): Int = columnNames.size

    override fun getColumnName(column: Int): String = columnNames[column]

    override fun getColumnClass(columnIndex: Int): Class<*> = when (columnIndex) {
        0 -> java.lang.Boolean::class.java
        else -> String::class.java
    }

    override fun isCellEditable(rowIndex: Int, columnIndex: Int): Boolean = 
        columnIndex == 0 || columnIndex == 2 || columnIndex == 5

    override fun getValueAt(rowIndex: Int, columnIndex: Int): Any {
        val attack = displayedAttacks.getOrNull(rowIndex) ?: return ""
        return when (columnIndex) {
            0 -> attack.id in selectedIds
            1 -> DATE_FORMATTER.format(Instant.ofEpochMilli(attack.timestamp))
            2 -> attack.category
            3 -> attack.method
            4 -> attack.url
            5 -> attack.status
            6 -> attack.testerName
            else -> ""
        }
    }

    override fun setValueAt(aValue: Any?, rowIndex: Int, columnIndex: Int) {
        if (columnIndex == 0 && aValue is Boolean) {
            val attack = displayedAttacks.getOrNull(rowIndex) ?: return
            if (aValue) {
                selectedIds.add(attack.id)
            } else {
                selectedIds.remove(attack.id)
            }
            fireTableCellUpdated(rowIndex, columnIndex)
        } else if (columnIndex == 2 || columnIndex == 5) {
            val attack = displayedAttacks.getOrNull(rowIndex) ?: return
            val valStr = aValue as? String ?: return
            
            var updatedEntry = attack
            if (columnIndex == 2) { // Category
                updatedEntry = attack.copy(category = valStr)
            } else if (columnIndex == 5) { // Status
                updatedEntry = attack.copy(status = valStr)
            }
            
            // Persist changes
            storageManager.updateAttack(updatedEntry)
            
            // Update local memory
            val allIndex = allAttacks.indexOfFirst { it.id == attack.id }
            if (allIndex != -1) {
                allAttacks[allIndex] = updatedEntry
            }
            refreshDisplay()
        }
    }

    fun setWrappedSelected(id: String, selected: Boolean) {
        if (selected) selectedIds.add(id) else selectedIds.remove(id)
        fireTableDataChanged()
    }

    companion object {
        private val DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
            .withZone(ZoneId.systemDefault())
    }
}
